### DISCUSSION TITLE
Can we remove the dependence of the DB on jagged arrays?

### DISCUSSION THREAD

**Discussion by opotowsky** [terrapower/armi#1669] [General]
Our dependence on jagged arrays means we have a max numpy version of 1.23.5. 

What is the basis for needing jagged arrays? Can we restructure something to remove this?

We have a ticket open for this [here](https://github.com/terrapower/armi/issues/1036).

@john-science @onufer @albeanth @mgjarrett @jakehader @sombrereau

**Comment by albeanth** [comment]
Do you know where we use jagged arrays?

**Reply by jakehader** [reply]
I'm not entirely sure the basis, but I have to assume it's to reduce DB size with zero values / sparse data in an array / matrix (e.g., Nones, NaNs)? @youngmit might come to our rescue here 😉. If you search "jagged" in the repository there are several instances in the DB implementation docs and comments @albeanth.

**Comment by mgjarrett** [comment]
There aren't actually any jagged arrays stored in the database. But `Database3` sometimes creates a jagged numpy array before flattening it into a 1D array and storing that 1D array (along with offsets) in the database.

So we would probably just need to build our own `JaggedArray` class that would be a special type of `list`. You would create a `JaggedArray` instance, append a bunch of numpy arrays to the list, and then concatenate the whole thing into a flattened 1D numpy array with offsets and spit that back out to be dumped into the database.

**Reply by mgjarrett** [reply]
I'm looking more closely, and it appears that we also create jagged numpy arrays for convenience when reading the data back out of the HDF5.

https://github.com/terrapower/armi/blob/827d223cd6fc085c039390a2c3a0391f4e611fe5/armi/bookkeeping/db/database3.py#L1361-L1363

It would take some work to refactor all of this to not require a jagged numpy array, but it's probably not _too_ complicated.

**Reply by john-science** [reply]
The problem I originally had in [this ticket](https://github.com/terrapower/armi/issues/1036) was identifying, conclusively, all the locations we use jagged arrays.

That would make this work significantly easier.

**Comment by opotowsky** [comment]
Ok this isn't as urgent as I thought. I tried to recreate the errors I thought I had on linux with numpy 1.23.5 and am unable to do so. 

It's still a good idea to put into a longer term plan so we aren't locked into ever-aging versions of numpy, but instead of a this-year thing it's a next year thing ;-)

**Comment by john-science** [comment]
I would like to note that because we can't use the latest versions of NumPy (due specifically to jagged arrays), ARMI cannot support Python 3.12: https://github.com/terrapower/armi/issues/246

That's a pretty big deal.

**Reply by albeanth** [reply]
@john-science do you think this should be apart of ARMI v0.4?

**Reply by john-science** [reply]
I very much do, yes.  I think I would have to fight pretty hard to get that put on the calendar though.

**Comment by ntouran** [comment]
Jagged arrays are used to store pin detailed parameter data on assemblies that have differing numbers of pins. So if you have a CR with 17 pins and a fuel assembly with 127 pins, that ends up using jagged arrays. We can definitely refactor the way these data are stored in the DB in order to support the deprecation, but being able to store pin detailed param data on multi-num pin cores is an essential feature for the framework.

**Reply by john-science** [reply]
Check out Michael's PR on the topic: https://github.com/terrapower/armi/pull/1726

**Comment by mgjarrett** [comment]
I was able to get something working a couple of months ago: https://github.com/terrapower/armi/tree/jaggedArrays but then interest died down so I put it on the back burner.

It's not a finished product yet, but I think it's enough to show notionally that this can be done without a huge overhaul. I basically just added a new class to database3.py. I'd definitely like to have more eyes on it and spend some more time testing corner cases and making sure everything is implemented efficiently.

There are a couple unit tests that are passing on my local machine with the old version of numpy, but fail when running with numpy 1.26.4 on GitHub actions. I think these failures could probably be resolved pretty easily, but I haven't gotten around to it yet.

**Reply by opotowsky** [reply]
🦸‍♂️

**Comment by john-science** [comment]
We have [a ticket](https://github.com/terrapower/armi/issues/246) open for this, and [a PR](https://github.com/terrapower/armi/pull/1726). 

So, we are definitely doing this soon, and I will close this discussion.
